backoff.c

```diff
@@ -0,0 +1,87 @@
+#include "common.h"
+
+#include <ext/standard/php_rand.h>
+#include <ext/standard/php_mt_rand.h>
+
+#include "backoff.h"
+
+zend_long random_range(zend_long min, zend_long max) {
+    if (max <= min) {
+        max = min + 1;
+    }
+
+  return php_mt_rand_range(min, max);
+}
+
+zend_long default_backoff(struct Backoff *self, int retry_index) {
+    zend_long backoff = retry_index ? self->base : random_range(0, self->base);
+    return MIN(self->cap, backoff);
+}
+
+zend_long constant_backoff(struct Backoff *self, int retry_index) {
+    zend_long backoff = self->base;
+    return MIN(self->cap, backoff);
+}
+
+zend_long uniform_backoff(struct Backoff *self, int retry_index) {
+  zend_long backoff = random_range(0, self->base);
+    return MIN(self->cap, backoff);
+}
+
+zend_long exponential_backoff(struct Backoff *self, int retry_index) {
+    zend_long pow = MIN(retry_index, 10);
+    zend_long backoff = self->base * (1 << pow);
+    return MIN(self->cap, backoff);
+}
+
+zend_long full_jitter_backoff(struct Backoff *self, int retry_index) {
+    zend_long pow = MIN(retry_index, 10);
+    zend_long backoff = self->base * (1 << pow);
+    zend_long cap = MIN(self->cap, backoff);
+    return random_range(0, self->cap);
+}
+
+zend_long equal_jitter_backoff(struct Backoff *self, int retry_index) {
+    zend_long pow = MIN(retry_index, 10);
+    zend_long backoff = self->base * (1 << pow);
+    zend_long temp = MIN(self->cap, backoff) >> 1;
+    return temp + random_range(0, temp);
+}
+
+zend_long decorrelated_jitter_backoff(struct Backoff *self, int retry_index) {
+    zend_long max_backoff = MAX(self->base, self->previous_backoff * 3);
+    zend_long temp = random_range(self->base, max_backoff);
+    self->previous_backoff = MIN(self->cap, temp);
+    return self->previous_backoff;
```

Comment:
```suggestion
    zend_long next_max = self->previous_backoff * 3;
    if (self->base < next_max)
        self->previous_backoff = random_range(self->base, next_max);
    else
        self->previous_backoff = random_range(next_max, self->base);

    return MIN(self->cap, self->previous_backoff);

```

If I've misunderstood this algorithm from the blog post let me know, but I _think_ this is more consistent with the formula:

```c
sleep = min(cap, random_range(base, sleep * 3))
```

The current implementation always picks `self->base + 1` for the first iteration because `MAX(base, previous) always initially returns`base` (because previous is 0).

That makes `random_range(base, max_backoff)` == `random_range(base, base)`.

Reply:
You are right. Fixed.